### PR TITLE
Fixed several CSP issues requiring 'unsafe-inline' rule.

### PR DESCRIPTION
**Pull request by EpicVoyage**
As mentioned on #2055, this patches the features I use so that they play well with a stricter Content-Security-Policy (CSP). It does not fix everything, but should serve as a good template to help others get started on fixing other features.

My configuration: `Content-Security-Policy: base-uri 'self'; default-src mydomain.com 'nonce-{{nonce}}'`

The changes add support for [Angular's Nonce support](https://angular.io/api/core/CSP_NONCE). `<style>` tags inserted into the DOM will contain this Nonce, if one is configured.

This also removes a lot of inline `style` attributes from HTML tags, which are not permitted due to scripting possibilities.

It removes several uses of `innerHTML`. This has been replaced a few different ways - setting text instead of HTML or using Renderer2's `appendChild()` method. There are more of these references left - my apologies.

A quick sidenote for anyone doing Windows development! You may need to pull the published `pdf.js` files in order to provide minified and unminified scripts. The easiest way on Windows is to do this:

    npm pack ngx-extended-pdf-viewer@19.4.1
    tar -xzvf ngx-extended-pdf-viewer-19.4.1.tgz --strip-components=2 -C projects\ngx-extended-pdf-viewer\assets package/assets

### CODE REVIEW

**Comment by stephanrauh** [comment]
Whow. I'm impressed. That was a lot of work, and nicely done. You even met my subconscious naming conventions. Chris, when we meet in person, remind me I owe you a beer! Or a coffee, or whatever you like. :)

I've left two comments, but both of them are complaining on a high level. The only reason I don't merge your PR right now is it's late in the evening. Consider it merged!

**Comment by EpicVoyage** [comment]
That sounds good, Stephan. Glad you like the updates!

I flipped through the merge details and did not see those 2 comments. They did not appear to be in SonarCloud either? There was a typo in my code though, which has now been fixed.

**Comment by stephanrauh** [comment]
I wasn't aware I have to submit a review before you can see it. Now you can see the review (see above).

BTW, thanks for checking Sonarcloud, too! And thanks for fixing the typo - I missed it yesterday. Keep on the good work!

File: projects/ngx-extended-pdf-viewer/src/lib/sidebar/pdf-sidebar/pdf-sidebar-content/pdf-sidebar-content.component.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ export class PdfSidebarContentComponent implements OnDestroy {
     return `${top}px`;
   }
 
-  constructor(private pdfCspPolicyService: PdfCspPolicyService) {
+  constructor(private pdfCspPolicyService: PdfCspPolicyService, private renderer: Renderer2) {
```
[the remark this answers is not in the document]
Comment (EpicVoyage) [reply projects/ngx-extended-pdf-viewer/src/lib/sidebar/pdf-sidebar/pdf-sidebar-content/pdf-sidebar-content.component.ts:54]: You are right, `renderer` is not used in this class after all. I will revert that change.

File: projects/ngx-extended-pdf-viewer/src/lib/toolbar/pdf-context-menu/pdf-context-menu.component.html [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <!-- the context menu is deactivated because only Firefox supports it -->
-<div style="display:none" type="context" id="viewerContextMenu">
-  <div style="display:none" id="contextFirstPage"></div>
-  <div style="display:none" id="contextLastPage"></div>
-  <div style="display:none" id="contextPageRotateCw"></div>
-  <div style="display:none" id="contextPageRotateCcw"></div>
+<div class="display-none" type="context" id="viewerContextMenu">
```
[the remark this answers is not in the document]
Comment (EpicVoyage) [reply projects/ngx-extended-pdf-viewer/src/lib/toolbar/pdf-context-menu/pdf-context-menu.component.html:2]: Yes, let's use that.

**Comment by EpicVoyage** [comment]
You raise valid points in review. I have:

- Cleaned up the two issues you raised.
- Made SVG size definitions consistent for this patch, instead of splitting them between `width`/`height` attributes and CSS styles.
- Made `NgxExtendedPdfViewerService` Injectable because the compiler asked for this when I re-tested the changes locally.

Hopefully that puts this commit in a good place! Thanks for keeping this project top-notch.

**Comment by stephanrauh** [comment]
Hm... I've got a question. This combination of classes is interesting:
```html
class="hidden" [class.hidden]="hideSidebarToolbar">
```
I've used something similar in the past, too, but it felt like a hack. Is it a hack? Or is it acceptable?

**Comment by stephanrauh** [comment]
After mergin, a very unexpected error popped up in this method:
```ts
  private createElementFromHTML(htmlString): HTMLElement {
    const div = document.createElement('div');
    this.pdfCspPolicyService.addTrustedHTML(div, htmlString);
    return div.firstChild as HTMLElement;
  }
```
It used to return a DOM tree. However, now it's not a tree, but only the top-level node, and everything inside is escaped:
image

It's possibly - or most likely - unrelated to your change, but it still works in the showcase, and there's almost no change since then except accepting your PR... so what's gone wrong?

**Comment by EpicVoyage** [comment]
I will take a look when I sit down at a computer this evening.

You have me questioning whether it is necessary to specify the class name to also use `[class.hidden]=`. I will check that also.

**Comment by stephanrauh** [comment]
> You have me questioning whether it is necessary to specify the class name to also use [class.hidden]=. I will check that also.

More to the point, I wonder if that‘s correct Angular code. It works, but is it future-proof?

**Comment by EpicVoyage** [comment]
> More to the point, I wonder if that‘s correct Angular code. It works, but is it future-proof?

This is a currently-documented feature:

https://angular.io/guide/class-binding

Hopefully it will be around for a good, long time! It looks like we should not have to declare the class and then set the condition. I will clean that up.

**Comment by stephanrauh** [comment]
I've found and understood the thumbnails bug. I'm sure I can fix this soon.

I've also pushed a commit fixing a bug of the NgxExtendedPdfViewerService. For some reason unknown, `Render2` can only be injected into components, not in services.

**Comment by EpicVoyage** [comment]
I am very sorry for causing you extra work. It looks like my last round of testing was still done against my changes to `19.2.2` because I missed a package version change in one of my local build commands. That was before you had even published `PdfCspPolicyService`. I'll take all the blame for breaking things.

I have a fix for `createElementFromHTML()`, but need to retest several changes before sending it your way.

**Comment by stephanrauh** [comment]
Don't be sorry! I'm glad you've managed to fix the CSP issue. I considered it a bit embarrassing, but I couldn't wrap my head around it. It's perfectly normal that minor issues pop up when someone else starts testing.

I'm looking forward to your solution. ChatGPT suggested this approach:
```ts
export class AppComponent  {
  // not applicable: @ViewChild('targetDiv', { read: ViewContainerRef }) targetDiv!: ViewContainerRef;

  constructor(private renderer: Renderer2) { }

  // Method to modify and add the template content
  createThumbnail(..., container: HTMLDivElement) {
    const view = this.thumbnailViewTemplate.createEmbeddedView(null);

    const widthOfRing = `${pdfThumbnailView.canvasWidth + borderAdjustment}px`;
    const heightOfRing = `${pdfThumbnailView.canvasHeight + borderAdjustment}px`;

    // TODO: replace the placeholders WIDTH_OF_RING, HEIGHT_OF_RING, and PAGE_NUMBER
    // Caveat: we don't know which template the user defines, so these attributes can be anywhere

    // Attach the view to the target div

    container.insert(view); // TODO: is this possible? container is not a ViewContainerRef, but a HTMLDivElement, because it comes from pdf.js. If need be, we can change that
  }
}
```

**Comment by stephanrauh** [comment]
You improvement has landed with version 19.5.0. Now it also renders thumbnails, and I've even been able to remove some legacy code.

Enjoy!
Stephan

**Pull request merged**